In [ ]:
import pandas as pd
from datetime import datetime
from scipy.stats import ttest_ind
import matplotlib.pyplot as plt
import numpy as np


In [ ]:
data = pd.read_csv("/content/daily_btc_1.csv")

**DATA PRE-PROCESSING**

In [ ]:
# Load the data again
btc_df = pd.read_csv("/content/daily_btc_1.csv")

# Convert 'Date' to datetime format and price columns to numeric format
btc_df['Date'] = pd.to_datetime(btc_df['Date'])
for col in ['Open', 'High', 'Low', 'Close']:
    btc_df[col] = btc_df[col].str.replace(',', '').astype(float)




In [ ]:
# Sort the data by date in ascending order
btc_df = btc_df.sort_values('Date')

# Define the date of the WHO COVID-19 pandemic declaration
covid_declaration_date = pd.to_datetime('2020-03-11')

# Split the data into before and after the declaration
btc_before_covid = btc_df[btc_df['Date'] < covid_declaration_date]
btc_after_covid = btc_df[btc_df['Date'] >= covid_declaration_date]

# Get descriptive statistics for the entire dataset, before COVID, and after COVID
desc_all = btc_df.describe()
desc_before = btc_before_covid.describe()
desc_after = btc_after_covid.describe()

desc_all, desc_before, desc_after

(               Open          High           Low         Close
 count   2015.000000   2015.000000   2015.000000   2015.000000
 mean   19240.985112  19744.907196  18694.158809  19246.125558
 std    16625.638414  17049.008242  16135.351689  16616.485889
 min      608.000000    612.000000    608.000000    608.000000
 25%     6584.500000   6751.500000   6437.500000   6591.000000
 50%    11428.000000  11711.000000  11056.000000  11397.000000
 75%    29504.500000  30318.000000  28941.500000  29502.500000
 max    67802.000000  68925.000000  66112.000000  67802.000000,
                Open          High           Low         Close
 count    945.000000    945.000000    945.000000    945.000000
 mean    6207.893122   6406.632804   5988.457143   6209.532275
 std     3698.164520   3858.480554   3503.270787   3692.145745
 min      608.000000    612.000000    608.000000    608.000000
 25%     3545.000000   3608.000000   3447.000000   3563.000000
 50%     6498.000000   6623.000000   6402.000000   650

In [ ]:
# Calculate the mean closing price for each group
mean_close_before = btc_before_covid['Close'].mean()
mean_close_after = btc_after_covid['Close'].mean()

# Conduct a t-test to compare the means
t_stat, p_value = stats.ttest_ind(btc_before_covid['Close'], btc_after_covid['Close'])

mean_close_before, mean_close_after, t_stat, p_value


(6209.532275132276, 30759.752336448597, -48.994267301079155, 0.0)

**Splitting data into Pre-Covid and Post-Covid**

In [ ]:
# Define the date of the WHO COVID-19 pandemic declaration
covid_declaration_date = pd.to_datetime('2020-03-11')

# Split the data into before and after the declaration
btc_before_covid = btc_df[btc_df['Date'] < covid_declaration_date]
btc_after_covid = btc_df[btc_df['Date'] >= covid_declaration_date]

btc_before_covid, btc_after_covid


(           Date    Open    High     Low   Close
 2014 2016-10-03   613.0   614.0   608.0   610.0
 2013 2016-10-04   610.0   613.0   610.0   612.0
 2012 2016-10-05   612.0   612.0   608.0   608.0
 2011 2016-10-06   608.0   613.0   608.0   612.0
 2010 2016-10-07   612.0   612.0   610.0   616.0
 ...         ...     ...     ...     ...     ...
 1084 2020-02-25  9607.0  9673.0  9262.0  9383.0
 1078 2020-02-26  9391.0  9425.0  8648.0  8763.0
 1072 2020-02-27  8754.0  8971.0  8535.0  8756.0
 1066 2020-02-28  8760.0  8903.0  8446.0  8676.0
 1062 2020-02-29  8676.0  8814.0  8596.0  8640.0
 
 [945 rows x 5 columns],
            Date     Open     High      Low    Close
 1059 2020-03-13   5747.0   6029.0   3966.0   5435.0
 1053 2020-03-14   5436.0   5766.0   5224.0   5354.0
 1047 2020-03-15   5354.0   5583.0   5074.0   5187.0
 1041 2020-03-16   5188.0   5959.0   4461.0   4904.0
 1035 2020-03-17   4904.0   5502.0   4882.0   5474.0
 ...         ...      ...      ...      ...      ...
 1294 2023-07-

In [ ]:
# Save the pre-COVID and post-COVID data to CSV files
btc_before_covid.to_csv("/content/pre_covid_data.csv", index=False)
btc_after_covid.to_csv("/content/post_covid_data.csv", index=False)


In [ ]:
# Get descriptive statistics for the entire dataset, before COVID, and after COVID
desc_all = btc_df['Close'].describe()
desc_before = btc_before_covid['Close'].describe()
desc_after = btc_after_covid['Close'].describe()

desc_all, desc_before, desc_after


(count     2015.000000
 mean     19246.125558
 std      16616.485889
 min        608.000000
 25%       6591.000000
 50%      11397.000000
 75%      29502.500000
 max      67802.000000
 Name: Close, dtype: float64,
 count      945.000000
 mean      6209.532275
 std       3692.145745
 min        608.000000
 25%       3563.000000
 50%       6501.000000
 75%       8589.000000
 max      19283.000000
 Name: Close, dtype: float64,
 count     1070.000000
 mean     30759.752336
 std      15007.442663
 min       4904.000000
 25%      19422.750000
 50%      28300.500000
 75%      42072.750000
 max      67802.000000
 Name: Close, dtype: float64)

In [ ]:
import scipy.stats as stats


# Perform a two-sample t-test
t_statistic, p_value = stats.ttest_ind(pre_covid_price, post_covid_price, equal_var=False)

# Print the results
print(f"t-statistic: {t_statistic}")
print(f"p-value: {p_value}")



t-statistic: -51.76625265514535
p-value: 1.582987164371413e-309


The t-statistic is -51.77, which is much greater than the critical value of 1.96 for a two-tailed test with a significance level of 0.05. This means that we can reject the null hypothesis that the mean close price of Bitcoin is the same in the pre-COVID and post-COVID periods.

The p-value is 1.582987164371413e-309, which is essentially zero. This means that there is a zero probability that the difference in the mean close price of Bitcoin between the pre-COVID and post-COVID periods is due to chance.

In conclusion, the results of the two-sample t-test suggest that the mean close price of Bitcoin is significantly different in the pre-COVID and post-COVID periods. This means that the COVID pandemic had a significant impact on the price of Bitcoin.

In [ ]:
import statsmodels.tsa.arima.model as smt

# Drop any NaN values from the data
pre_covid_data_no_nan = pre_covid_data.dropna()
post_covid_data_no_nan = post_covid_data.dropna()

# Fit an ARIMA model on the pre-COVID data
model_pre = smt.ARIMA(pre_covid_data_no_nan['Close'].astype('float64').to_numpy(), order=(1,1,1))
model_pre_fit = model_pre.fit()

# Calculate the BIC value, MSE value, RMSE value, and AIC value for the pre-COVID ARIMA model
bic_pre = model_pre_fit.bic
mse_pre = np.mean((model_pre_fit.predict() - pre_covid_data_no_nan['Close'].astype('float64')**2))
rmse_pre = np.sqrt(mse_pre)
aic_pre = model_pre_fit.aic

# Fit an ARIMA model on the post-COVID data
model_post = smt.ARIMA(post_covid_data_no_nan['Close'].astype('float64').to_numpy(), order=(1,1,1))
model_post_fit = model_post.fit()

# Calculate the BIC value, MSE value, RMSE value, and AIC value for the post-COVID ARIMA model
bic_post = model_post_fit.bic
mse_post = np.mean((model_post_fit.predict() - post_covid_data_no_nan['Close'].astype('float64')**2))
rmse_post = np.sqrt(mse_post)
aic_post = model_post_fit.aic

# Print the results
print("The BIC value for the pre-COVID ARIMA model is:", bic_pre)
print("The MSE value for the pre-COVID ARIMA model is:", mse_pre)
print("The RMSE value for the pre-COVID ARIMA model is:", rmse_pre)
print("The AIC value for the pre-COVID ARIMA model is:", aic_pre)
print("The BIC value for the post-COVID ARIMA model is:", bic_post)
print("The MSE value for the post-COVID ARIMA model is:", mse_post)
print("The RMSE value for the post-COVID ARIMA model is:", rmse_post)
print("The AIC value for the post-COVID ARIMA model is:", aic_post)

The BIC value for the pre-COVID ARIMA model is: 14008.942074637622
The MSE value for the pre-COVID ARIMA model is: -52169605.33791222
The RMSE value for the pre-COVID ARIMA model is: nan
The AIC value for the pre-COVID ARIMA model is: 13994.391696139186
The BIC value for the post-COVID ARIMA model is: 18379.059540932194
The MSE value for the post-COVID ARIMA model is: -1171144479.5297284
The RMSE value for the post-COVID ARIMA model is: nan
The AIC value for the post-COVID ARIMA model is: 18364.13610419912


<ipython-input-21-e19cb89cb5d7>:14: RuntimeWarning: invalid value encountered in sqrt
  rmse_pre = np.sqrt(mse_pre)
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/sarimax.py:966: UserWarning: Non-stationary starting autoregressive parameters found. Using zeros as starting parameters.
  warn('Non-stationary starting autoregressive parameters'
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/sarimax.py:978: UserWarning: Non-invertible starting MA parameters found. Using zeros as starting parameters.
  warn('Non-invertible starting MA parameters found.'
<ipython-input-21-e19cb89cb5d7>:24: RuntimeWarning: invalid value encountered in sqrt
  rmse_post = np.sqrt(mse_post)


The code attempts to fit ARIMA models to pre-COVID and post-COVID Bitcoin price data. However, the code runs into an error because the MSE values for both models are negative. This means that the error terms for the models are larger than the actual values, which is not possible.


In [ ]:
#EXTREME GRADIENT BOOST MODEL
# Create the COVID-19 feature
btc_df['COVID'] = btc_df['Date'].apply(lambda x: 1 if x >= datetime(2020, 3, 11) else 0)

btc_df.head()

Date   Open   High    Low  Close  COVID
2014 2016-10-03  613.0  614.0  608.0  610.0      0
2013 2016-10-04  610.0  613.0  610.0  612.0      0
2012 2016-10-05  612.0  612.0  608.0  608.0      0
2011 2016-10-06  608.0  613.0  608.0  612.0      0
2010 2016-10-07  612.0  612.0  610.0  616.0      0

In [ ]:
# Check for missing values in the dataset
btc_df.isnull().sum()


Date     0
Open     0
High     0
Low      0
Close    0
COVID    0
dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Drop rows with missing values
btc_df_clean = btc_df.dropna()

# Define X and y
X = btc_df_clean.drop(['Date', 'Close'], axis=1)
y = btc_df_clean['Close']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the model
model = XGBRegressor()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mse, mae, r2


(413696.5077608406, 356.1109276859104, 0.9985674717664168)